relay-sampling/src/evaluation.rs

```diff
@@ -70,164 +67,229 @@ fn check_unsupported_rules(
             if processing_enabled {
                 relay_log::error!("found unsupported sampling mode even as processing Relay");
             }
-
-            return None;
+            return SamplingResult::NoMatch;
+        }
+        None => {
+            relay_log::info!("cannot sample without at least one sampling config");
+            return SamplingResult::NoMatch;
         }
     };
-    match_result.set_sample_rate(sample_rate);
 
-    // Only if we arrive at this stage, it means that we have found a match and we want to prepare
-    // the data for making the sampling decision.
-    Some(match_result)
+    // We perform the rule matching with the multi-matching logic on the merged rules.
+    let Some(rules) =
+        verify_and_merge_rules(processing_enabled, sampling_config, root_sampling_config)
+    else {
+        return SamplingResult::NoMatch;
+    };
+
+    get_sampling_match(rules, event, dsc, now, adjust_sample_rate)
 }
 
 /// Represents the specification for sampling an incoming event.
-#[derive(Clone, Debug, PartialEq, Serialize)]
-pub struct SamplingMatch {
-    /// The sample rate to use for the incoming event.
-    pub sample_rate: f64,
+#[derive(Default, Clone, Debug, PartialEq)]
+pub enum SamplingResult {
+    /// The event matched a sampling condition.
+    Match {
+        /// The sample rate to use for the incoming event.
+        sample_rate: f64,
+        /// The seed to feed to the random number generator which allows the same number to be
+        /// generated given the same seed.
+        ///
+        /// This is especially important for trace sampling, even though we can have inconsistent
+        /// traces due to multi-matching.
+        seed: Uuid,
+        /// The list of rule ids that have matched the incoming event and/or dynamic sampling context.
+        matched_rules: MatchedRuleIds,
+        /// Whether this sampling match results in the item getting sampled.
+        /// It's essentially a cache, as the value can be deterministically derived from
+        /// the sample rate and the seed.
+        is_kept: bool,
+    },
+    /// The event did not match a sampling condition.
+    #[default]
+    NoMatch,
+}
 
-    /// The seed to feed to the random number generator which allows the same number to be
-    /// generated given the same seed.
-    ///
-    /// This is especially important for trace sampling, even though we can have inconsistent
-    /// traces due to multi-matching.
-    pub seed: Uuid,
+impl SamplingResult {
+    /// Returns the sample rate.
+    pub fn sample_rate(&self) -> Option<f64> {
+        if let Self::Match { sample_rate, .. } = self {
+            return Some(*sample_rate);
+        }
+        None
+    }
 
-    /// The list of rule ids that have matched the incoming event and/or dynamic sampling context.
-    pub matched_rule_ids: MatchedRuleIds,
-}
+    /// Returns true if the event matched on any rules.
+    pub fn no_match(&self) -> bool {
```

Comment:
```suggestion
    pub fn is_no_match(&self) -> bool {
```

And consider to introduce parity via `is_match`.